// @cpanel/api/dist/http/verb.js
// MIT License
//
// Copyright 2021 cPanel L.L.C.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
//  of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the
// rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
// sell copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
// DEALINGS IN THE SOFTWARE.
/**
 * Common http verbs
 */
export var HttpVerb;
(function (HttpVerb) {
    /**
     * Get request
     */
    HttpVerb[HttpVerb["GET"] = 0] = "GET";
    /**
     * Head request
     */
    HttpVerb[HttpVerb["HEAD"] = 1] = "HEAD";
    /**
     * Post request
     */
    HttpVerb[HttpVerb["POST"] = 2] = "POST";
    /**
     * Put request
     */
    HttpVerb[HttpVerb["PUT"] = 3] = "PUT";
    /**
     * Delete request
     */
    HttpVerb[HttpVerb["DELETE"] = 4] = "DELETE";
    /**
     * Connect request
     */
    HttpVerb[HttpVerb["CONNECT"] = 5] = "CONNECT";
    /**
     * Options request
     */
    HttpVerb[HttpVerb["OPTIONS"] = 6] = "OPTIONS";
    /**
     * Trace request
     */
    HttpVerb[HttpVerb["TRACE"] = 7] = "TRACE";
    /**
     * Patch request
     */
    HttpVerb[HttpVerb["PATCH"] = 8] = "PATCH";
})(HttpVerb || (HttpVerb = {}));
